### PR TITLE
add python 3.8, 3.9 support and migration to pytest

### LINKED ISSUE #3046: python 3.8 support
Would be great if Luigi officially supported python >3.7, this issue is specifically for py3.8.

Related: 
* https://github.com/spotify/luigi/issues/2886
* https://github.com/spotify/luigi/issues/2939, @tirkarthi have you been using luigi on python 3.9? What's your experience?

### PR DESCRIPTION
**Pull request by hirosassa**
## Description

fix  #3046, related to #2886, #2939

This PR migrates from nose to pytest and adds python 3.8 and 3.9 support.

Details are as follows
- migrate from `nose` to `pytest`
  - place `test/conftest.py` to change test root
  - replace `@attr` decorator into `pytest.mark` to group tests
  - fix a few imports e.g. [here](https://github.com/spotify/luigi/pull/3048/files#diff-e590b5303093e390593b16c301f3ad6e3728f6afb8029d0f2fcb16964274ed53R19)
  - attach namespace on some test tasks to isolate each test e.g. [here](https://github.com/spotify/luigi/pull/3048/files#diff-053b30ca6a42cf2e4792b6b8ee44adf480820b1ad01ecce3942ec7dc816ff0d1R45)
- add py38 and py39 CI on `tox` and `travis`
  - update [test dependencies](https://github.com/spotify/luigi/pull/3048/files#diff-ef2cef9f88b4fe09ca3082140e67f5ad34fb65fb6e228f119d3812261ae51449R46) 
  - update virtualenv to resolve [cryptography installation problem](https://travis-ci.community/t/pip-install-cryptography-fails-on-py36/11233) 
    - [diff](https://github.com/spotify/luigi/pull/3048/files#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R107)
  - replace hypothesis.extra.datetime into hypothesis.strategies because of it's deprecation 
    - [diff](https://github.com/spotify/luigi/pull/3048/files#diff-d602eb271306017d537b4e752eecddfd78d03bd496dc7fca0c00e63ed75ca7d0R34)

### pytest migration
[`nose`](https://github.com/nose-devs/nose)  is not in active development and several deprecation warnings are occurred.

### add python 3.8, 3.9 tests
easy to follow latest python versions

## Have you tested this? If so, how?
 

I ran tests locally and on travis, too

### CODE REVIEW

**Comment by hirosassa** [comment]
@honnix @dlstadther @Tarrasch 
cc @ravwojdyla 
Does anybody know how to run CI with py38 environment on travis?
Timeout on `tox` step will always be occurred.

https://travis-ci.org/github/spotify/luigi/jobs/762401923

(I'm sorry for very very dirty and many commits. I'll squash later.)

**Comment by ravwojdyla** [comment]
@hirosassa 🤔 is py38 also timing out locally?

**Comment by hirosassa** [comment]
@ravwojdyla No, This is just occurred only on travis.

**Comment by ravwojdyla** [comment]
@hirosassa interesting. Travis has limited support for debugging builds: https://docs.travis-ci.com/user/running-build-in-debug-mode/. But I imagine we can't use that here, in the meantime we could also add `-vvvv` to tox commands in `./scripts/ci/conditional_tox.sh` to hopefully see more output and maybe the cause?

**Comment by hirosassa** [comment]
@ravwojdyla Thank you! I'll try `-vvv` option.

**Comment by hirosassa** [comment]
@ravwojdyla The cause of timeout is resolving dependencies on python 3.8 and I fixed it. Thanks!
I'll struggle against some failure tests.
https://travis-ci.org/github/spotify/luigi/builds/762528137

**Comment by ravwojdyla** [comment]
@hirosassa thanks for the update. Oh wow, there is quite a lot of failures on py38!

FYI @lallea since you have mentioned you've been using luigi on py38 in https://github.com/spotify/luigi/issues/3046#issuecomment-781431232. There seem to be quite a number of test failures, hopefully that doesn't affect your workloads, just wanted to let you know.

**Comment by ravwojdyla** [comment]
@hirosassa what do you think is the right way forward?

**Comment by lallea** [comment]
@ravwojdyla Thanks for the heads up. Lots of failures, indeed, and on core things. Surprising, since we have not seen any trouble. Works on our machines. :-)

**Comment by hirosassa** [comment]
@ravwojdyla @lallea I fixed almost all of the failures now (most of failures comes from coverage bug on python 3.8 and fixed by update it).

https://travis-ci.org/github/spotify/luigi/builds/762532477
https://travis-ci.org/github/spotify/luigi/jobs/762532486

**Comment by ravwojdyla** [comment]
That's a huge relief! Really appreciate your work @hirosassa!

**Comment by hirosassa** [comment]
@ravwojdyla @honnix @dlstadther 
Almost, all of the tests passed. Do you have any idea to solve following test's error?

```
test/scheduler_test.py::StableDoneCooldownSecsTest::test_sending_same_task_twice_without_cooldown_leads_to_double_run
```

<details>
<pre>
<code>
=================================== FAILURES ===================================

_ StableDoneCooldownSecsTest.test_sending_same_task_twice_without_cooldown_leads_to_double_run _

self = <scheduler_test.StableDoneCooldownSecsTest testMethod=test_sending_same_task_twice_without_cooldown_leads_to_double_run>

    @with_config({'scheduler': {'stable_done_cooldown_secs': '0'}})

    def test_sending_same_task_twice_without_cooldown_leads_to_double_run(self):

        second_run_result = self.get_second_run_result_on_double_run()

>       self.assertEqual(second_run_result.scheduling_succeeded, False)

E       AssertionError: True != False

test/scheduler_test.py:366: AssertionError

----------------------------- Captured stdout call -----------------------------

INFO: No prior state file exists at /var/lib/luigi-server/state.pickle. Starting with empty state

INFO: Scheduler starting up

INFO: logging already configured

DEBUG: Starting new HTTP connection (1): localhost:8082

INFO: 200 POST /api/ping (127.0.0.1) 2.42ms

DEBUG: http://localhost:8082 "POST /api/ping HTTP/1.1" 200 34

INFO: 200 POST /api/ping (127.0.0.1) 1.43ms

DEBUG: http://localhost:8082 "POST /api/ping HTTP/1.1" 200 34

INFO: 200 POST /api/add_task (127.0.0.1) 2.13ms

DEBUG: http://localhost:8082 "POST /api/add_task HTTP/1.1" 200 18

INFO: 200 POST /api/add_worker (127.0.0.1) 1.44ms

DEBUG: http://localhost:8082 "POST /api/add_worker HTTP/1.1" 200 18

INFO: 200 POST /api/get_work (127.0.0.1) 1.43ms

DEBUG: http://localhost:8082 "POST /api/get_work HTTP/1.1" 200 327

INFO: logging already configured

INFO: 200 POST /api/ping (127.0.0.1) 2.26ms

DEBUG: http://localhost:8082 "POST /api/ping HTTP/1.1" 200 34

DEBUG: Starting new HTTP connection (1): localhost:8082

INFO: 200 POST /api/ping (127.0.0.1) 1.39ms

DEBUG: http://localhost:8082 "POST /api/ping HTTP/1.1" 200 34

INFO: 200 POST /api/ping (127.0.0.1) 1.44ms

DEBUG: http://localhost:8082 "POST /api/ping HTTP/1.1" 200 34

INFO: 200 POST /api/add_task (127.0.0.1) 1.64ms

DEBUG: http://localhost:8082 "POST /api/add_task HTTP/1.1" 200 18

INFO: 200 POST /api/get_work (127.0.0.1) 1.32ms

DEBUG: http://localhost:8082 "POST /api/get_work HTTP/1.1" 200 154

INFO: 200 POST /api/ping (127.0.0.1) 1.40ms

DEBUG: http://localhost:8082 "POST /api/ping HTTP/1.1" 200 34

INFO: 200 POST /api/add_task (127.0.0.1) 1.59ms

DEBUG: http://localhost:8082 "POST /api/add_task HTTP/1.1" 200 18

INFO: 200 POST /api/add_worker (127.0.0.1) 1.47ms

DEBUG: http://localhost:8082 "POST /api/add_worker HTTP/1.1" 200 18

INFO: 200 POST /api/get_work (127.0.0.1) 1.44ms

DEBUG: http://localhost:8082 "POST /api/get_work HTTP/1.1" 200 154

INFO: Scheduler instance shutting down

WARNING: Failed saving scheduler state

Traceback (most recent call last):

  File "/home/travis/build/spotify/luigi/luigi/scheduler.py", line 445, in dump

    with open(self._state_path, 'wb') as fobj:

FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/luigi-server/state.pickle'

----------------------------- Captured stderr call -----------------------------
</code>
</pre>
</details>

**Comment by ravwojdyla** [comment]
@hirosassa great job!!

> left one same error for each python versions

So this smells like a regression from the package updates maybe? That test was introduced by https://github.com/spotify/luigi/commit/b4e44336ff4550a78e09e5aa03e1c2720fb3e985 as a fix for https://github.com/spotify/luigi/issues/2644: https://github.com/spotify/luigi/blob/7508cdf8af8161adfbe2d911e71f7aab38152a48/test/scheduler_test.py#L304-L366

has a number of knobs to tweak like the `time_to_check_secs`, `time_to_run_secs`, `parallel_scheduling_processes`. Maybe adding a couple of log messages between the steps would make it a bit more obvious why the task if failing. Also tweaking the knobs to more extreme values just for the debugging purposes?

**Comment by ravwojdyla** [comment]
@hirosassa 

> WARNING: Failed saving scheduler state
Traceback (most recent call last):
  File "/home/travis/build/spotify/luigi/luigi/scheduler.py", line 448, in dump
    with open(self._state_path, 'wb') as fobj:
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/luigi-server/state.pickle'

Does this directory: `/var/lib/luigi-server` exist at the time of the test run?

**Comment by hirosassa** [comment]
@ravwojdyla Thank you for your suggestion. I checked existence and writability of the directory by [this code](https://github.com/spotify/luigi/pull/3048/files#diff-ae5aa7d20825c4d1d2f7b4610ba0bcb9fd7700a84c22e14b5e2cfefcb047545bR445-R447), and found the following logs

```
INFO: state path: /var/lib/luigi-server/state.pickle, exists?: False, writable?: False
INFO: state path dir: /var/lib/luigi-server, exists?: False, writable?: False
INFO: state path root dir: /var/lib, exists?: True, writable?: False
```

However, there's no `mkdir` or preparation of the directories 🤔

**Comment by ravwojdyla** [comment]
@hirosassa that actually might be a red herring. So it seems like `test_sending_same_task_twice_without_cooldown_leads_to_double_run` is flaky on the travis CI only. Very weird. To limit this down, maybe we should try to run only a single environment instead of the test full matrix (essentially I wonder if there's some interaction there).

**Comment by hirosassa** [comment]
@ravwojdyla I found the reason `test_sending_same_task_twice_without_cooldown_leads_to_double_run` fails!
I removed coverage command in https://github.com/spotify/luigi/pull/3048/commits/644b445bcdb0e6d67fb8b230f31ebf0d3bb97543 , then all the core test passed.

- https://travis-ci.org/github/spotify/luigi/jobs/764230493
- https://travis-ci.org/github/spotify/luigi/jobs/764230497
- https://travis-ci.org/github/spotify/luigi/jobs/764230503
- https://travis-ci.org/github/spotify/luigi/jobs/764230507

**Comment by ravwojdyla** [comment]
@hirosassa great news! I wonder if this is a known issue(?) and what's the proper way forward?

**Comment by hirosassa** [comment]
@ravwojdyla finally! 🎉 
https://travis-ci.org/github/spotify/luigi/builds/764698712

**Comment by hirosassa** [comment]
@Tarrasch @honnix @dlstadther cc @ravwojdyla @lallea 
This PR is ready for review. Please check!

**Comment by ravwojdyla** [comment]
Great work @hirosassa!!

File: test/contrib/hdfs/webhdfs_client_test.py [hunk 1]
```diff
@@ -16,15 +16,18 @@
...
 
-from hdfs import InsecureClient
-from hdfs.ext.kerberos import KerberosClient
-from nose.plugins.attrib import attr
+import pytest
+try:
+    from hdfs import InsecureClient
+    from hdfs.ext.kerberos import KerberosClient
+except ImportError:
+    raise unittest.SkipTest('Unable to load hdfs module')
```
Comment (ravwojdyla) [inline test/contrib/hdfs/webhdfs_client_test.py:24]: You could probably also use [pytest.importorskip](https://docs.pytest.org/en/stable/reference.html#pytest-importorskip)?

File: test/customized_run_test.py [hunk 2]
```diff
@@ -27,6 +27,7 @@
 
 
 class DummyTask(luigi.Task):
+    task_namespace = 'customized_run'
```
Comment (ravwojdyla) [inline test/customized_run_test.py:30]: why is this now required? Is this related to path setup between nose and pytest?

File: tox.ini [hunk 3]
```diff
@@ -1,25 +1,41 @@
 [tox]
-envlist = py{35,36,37}-{cdh,hdp,core,contrib,apache,aws,gcloud,postgres,unixsocket,azureblob,dropbox}, visualiser, docs, flake8
+envlist = py{35,36,37,38,39}-{cdh,hdp,core,contrib,apache,aws,gcloud,postgres,unixsocket,azureblob,dropbox}, visualiser, docs, flake8
```
Comment (ravwojdyla) [inline tox.ini:2]: 🚀

File: tox.ini [hunk 4]
```diff
@@ -50,50 +66,45 @@ setenv =
...
 passenv = {[testenv]passenv}
 setenv =
     LC_ALL = en_US.utf-8
-    NOSE_EVAL_ATTR=not minicluster and not gcloud and not postgres and not unixsocket
     LUIGI_CONFIG_PATH={toxinidir}/test/testconfig/luigi.cfg
     TEST_VISUALISER=1
 commands =
     python --version
-    nosetests -v --tests=test/visualiser
+    pytest -m "not minicluster and not gcloud and not postgres and not unixsocket" test/visualiser
```
Comment (ravwojdyla) [inline tox.ini:107]: why these specific filters here?

**Review by ravwojdyla — APPROVED** [review]
@hirosassa looks great, and again great work figuring out all the weird errors!! I only have some questions inline, plus would suggest to open an issue to update test documentation to document `pytest` usage going forward.

File: tox.ini (hunk 4 above)
Comment (hirosassa) [reply tox.ini:107]: This filter is counter part of `NOSE_EVAL_ATTER` above. I just replaced it.

File: test/contrib/hdfs/webhdfs_client_test.py (hunk 1 above)
Comment (hirosassa) [reply test/contrib/hdfs/webhdfs_client_test.py]: Thank you for your suggestion! fixed in d4de5fe

File: test/customized_run_test.py (hunk 2 above)
Comment (hirosassa) [reply test/customized_run_test.py]: I added a comment in a542dee

**Comment by ravwojdyla** [comment]
👋 @lallea do you mind taking a look (review) as well (if you have time) please. I assume you would like to have this in as well :)

**Review by dlstadther — COMMENTED** [review]
Changes look good to me.

Could you also update the `Readme.md` to denote update to Python versions being tested and "supported"?

**Comment by hirosassa** [comment]
@dlstadther Thank you for the review! added in f7053d9

**Review by dlstadther — APPROVED** [review]

**Comment by lallea** [comment]
All looks good to me, with the reservation that I know little about pytest.

Thanks @hirosassa for putting down all this work!

**Comment by ravwojdyla** [comment]
👋   @honnix @narape what's the best way to request review/merge from the Spotify Team?

**Comment by narape** [comment]
:+1:  LGTM, @hirosassa Thanks for the contribution

**Comment by hirosassa** [comment]
Thanks for your comments and reviews! @ravwojdyla @lallea @dlstadther @narape

**Comment by ravwojdyla** [comment]
@hirosassa thanks for all you great work!

Btw, just noticed that it might be also worth updating the [setup.py](https://github.com/spotify/luigi/blob/master/setup.py#L113-L115) to include python 3.{8,9}. Let me know if that's something you would like to handle as well.

**Comment by hirosassa** [comment]
@ravwojdyla Oh, I missed it. I'll submit small PR to fix this.
